services/rpcfilters/logs_filter.go

```diff
@@ -0,0 +1,120 @@
+package rpcfilters
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type logsFilter struct {
+	mu                  sync.Mutex
+	logs                []types.Log
+	lastSeenBlockNumber uint64
+	lastSeenBlockHash   common.Hash
+
+	id    rpc.ID
+	crit  ethereum.FilterQuery
+	timer *time.Timer
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	done   chan struct{}
+}
+
+func (f *logsFilter) add(data interface{}) {
+	logs := data.([]types.Log)
+	filtered, num, hash := filterLogs(logs, f.crit, f.lastSeenBlockNumber, f.lastSeenBlockHash)
+	f.mu.Lock()
+	f.lastSeenBlockNumber = num
+	f.lastSeenBlockHash = hash
+	if len(filtered) > 0 {
```

Comment:
This `if` should not be needed: https://play.golang.org/p/E7ARiwj2CV0

Reply:
yeah, i had Lock inside the if before i added lastSeenBlockNumber and lastSeenBlockhash